File: src/main/services/geodb-downloader.ts
import { app } from "electron";
import path from "node:path";
import fs from "node:fs/promises";
import https from "node:https";
import { createWriteStream } from "node:fs";
import { createGunzip } from "node:zlib";
import { getLogger } from "./logger";
import { getSettingsStore } from "./persistence";
import { reloadGeoDatabases } from "./geo";
import type { GeoDbDownloadProgress } from "@common/ipc";

const log = getLogger();

type ProgressCallback = (progress: GeoDbDownloadProgress) => void;

const EDITIONS = ["GeoLite2-City", "GeoLite2-ASN"] as const;
type Edition = (typeof EDITIONS)[number];

function buildDownloadUrl(edition: Edition, licenseKey: string): string {
  return (
    `https://download.maxmind.com/app/geoip_download` +
    `?edition_id=${edition}&license_key=${encodeURIComponent(licenseKey)}&suffix=tar.gz`
  );
}

function getDatabaseDir(): string {
  return path.join(app.getPath("userData"), "databases");
}

function downloadFile(
  url: string,
  destPath: string,
  onPercent: (percent: number) => void,
  maxRedirects = 5
): Promise<void> {
  return new Promise((resolve, reject) => {
    const doRequest = (requestUrl: string, redirectsLeft: number) => {
      const parsed = new URL(requestUrl);
      log.info(`[geodb] HTTPS GET ${parsed.hostname}${parsed.pathname}${parsed.search.replace(/license_key=[^&]+/, "license_key=***")}`);

      const req = https.request(
        {
          hostname: parsed.hostname,
          port: 443,
          path: parsed.pathname + parsed.search,
          method: "GET",
          headers: {
            "User-Agent": "VisTracer/0.1 (GeoIP database downloader)"
          }
        },
        (res) => {
          log.info(`[geodb] HTTP ${res.statusCode}`);
          // Follow redirects (301, 302, 303, 307, 308)
          if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
            res.resume(); // Drain redirect response body to free the socket
            if (redirectsLeft <= 0) {
              reject(new Error("Too many redirects"));
              return;
            }
            doRequest(res.headers.location, redirectsLeft - 1);
            return;
          }

          if (!res.statusCode || res.statusCode >= 400) {
            let body = "";
            res.on("data", (chunk: Buffer) => { body += chunk.toString(); });
            res.on("end", () => {
              log.error(`[geodb] Download error: ${res.statusCode} ${body}`);
              reject(new Error(`MaxMind download failed (${res.statusCode}): ${body || res.statusMessage}`));
            });
            return;
          }

          const contentLength = Number(res.headers["content-length"] || 0);
          let downloaded = 0;
          const fileStream = createWriteStream(destPath);

          res.on("data", (chunk: Buffer) => {
            fileStream.write(chunk);
            downloaded += chunk.length;
            if (contentLength > 0) {
              onPercent(Math.round((downloaded / contentLength) * 100));
            }
          });

          res.on("end", () => {
            fileStream.end();
            fileStream.on("finish", resolve);
            fileStream.on("error", reject);
          });

          res.on("error", (err) => {
            fileStream.destroy();
            reject(err);
          });
        }
      );

      req.on("error", reject);
      req.end();
    };

    doRequest(url, maxRedirects);
  });
}

async function downloadAndExtract(
  edition: Edition,
  licenseKey: string,
  onProgress: ProgressCallback
): Promise<string> {
  const url = buildDownloadUrl(edition, licenseKey);
  log.info(`[geodb] Downloading ${edition} from ${url.replace(/license_key=[^&]+/, "license_key=***")}`);
  const dbDir = getDatabaseDir();
  await fs.mkdir(dbDir, { recursive: true });

  onProgress({ stage: "downloading", edition, percent: 0 });

  // Use Node's https module instead of the global fetch (which Electron overrides
  // with Chromium's network stack). Electron's fetch may reject valid license keys
  // due to cross-origin redirect handling differences.
  const tarGzPath = path.join(dbDir, `${edition}.tar.gz`);
  await downloadFile(url, tarGzPath, (percent) => {
    onProgress({ stage: "downloading", edition, percent });
  });

  onProgress({ stage: "extracting", edition });

  // Extract .mmdb from tarball
  // tar.gz contains: <edition>_<date>/<edition>.mmdb
  const mmdbPath = path.join(dbDir, `${edition}.mmdb`);
  await extractMmdbFromTarGz(tarGzPath, edition, mmdbPath);

  // Clean up tar.gz
  await fs.unlink(tarGzPath).catch(() => {});

  onProgress({ stage: "complete", edition, percent: 100 });
  return mmdbPath;
}

export async function extractMmdbFromTarGz(
  tarGzPath: string,
  edition: string,
  outputPath: string
): Promise<void> {
  // Simple tar parser — we only need to find the .mmdb file
  const { createReadStream } = await import("node:fs");
  const gunzip = createGunzip();
  const source = createReadStream(tarGzPath);

  const chunks: Buffer[] = [];
  const decompressed = source.pipe(gunzip);

  for await (const chunk of decompressed) {
    chunks.push(Buffer.from(chunk));
  }

  const tarBuffer = Buffer.concat(chunks);
  const mmdbFilename = `${edition}.mmdb`;
  let offset = 0;

  while (offset < tarBuffer.length - 512) {
    // Read tar header (512 bytes)
    const header = tarBuffer.subarray(offset, offset + 512);

    // Check for end-of-archive (two zero blocks)
    if (header.every((b) => b === 0)) break;

    // Extract filename from header (first 100 bytes, null-terminated)
    const nameEnd = header.indexOf(0, 0);
    const name = header.subarray(0, Math.min(nameEnd, 100)).toString("utf8");

    // Extract file size from header (octal, bytes 124-135)
    const sizeStr = header.subarray(124, 136).toString("utf8").trim();
    const size = parseInt(sizeStr, 8) || 0;

    offset += 512; // Move past header

    if (name.endsWith(mmdbFilename)) {
      const fileData = tarBuffer.subarray(offset, offset + size);
      await fs.writeFile(outputPath, fileData);
      return;
    }

    // Skip file data (padded to 512-byte boundary)
    offset += Math.ceil(size / 512) * 512;
  }

  throw new Error(`${mmdbFilename} not found in downloaded archive`);
}

export async function downloadGeoDatabases(
  licenseKey: string,
  onProgress: ProgressCallback
): Promise<{ cityPath: string; asnPath: string }> {
  const results: Record<string, string> = {};

  for (const edition of EDITIONS) {
    try {
      const dbPath = await downloadAndExtract(edition, licenseKey, onProgress);
      results[edition] = dbPath;
      log.info(`Downloaded ${edition} to ${dbPath}`);
    } catch (error) {
      const message = error instanceof Error ? error.message : "Download failed";
      log.error(`Failed to download ${edition}`, error);
      onProgress({ stage: "error", edition, error: message });
      throw error;
    }
  }

  const cityPath = results["GeoLite2-City"];
  const asnPath = results["GeoLite2-ASN"];

  // Update settings and reload readers
  const store = getSettingsStore();
  const geoSettings = store.get("geo");
  store.set("geo", {
    ...geoSettings,
    cityDbPath: cityPath,
    asnDbPath: asnPath,
    lastUpdated: Date.now()
  });

  await reloadGeoDatabases();

  return { cityPath, asnPath };
}

File: src-tauri/src/commands/geodb_downloader.rs
use crate::commands::geo::reload_geo_databases;
use crate::commands::geo::SharedGeoReaders;
use crate::commands::persistence::SharedStore;
use crate::types::GeoDbDownloadProgress;
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Emitter};

const EDITIONS: &[&str] = &["GeoLite2-City", "GeoLite2-ASN"];

fn build_download_url(edition: &str, license_key: &str) -> String {
    format!(
        "https://download.maxmind.com/app/geoip_download?edition_id={}&license_key={}&suffix=tar.gz",
        edition,
        urlencoding::encode(license_key)
    )
}

fn emit_progress(app: &AppHandle, progress: &GeoDbDownloadProgress) {
    if let Err(e) = app.emit("vistracer:geo:download-progress", progress) {
        log::warn!("Failed to emit download progress: {}", e);
    }
}

async fn download_file(url: &str, dest: &Path) -> Result<(), String> {
    let client = reqwest::Client::new();
    let resp = client
        .get(url)
        .header(
            "User-Agent",
            "VisTracer/0.2 (GeoIP database downloader)",
        )
        .send()
        .await
        .map_err(|e| format!("Download failed: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!(
            "MaxMind download failed ({}): {}",
            status,
            body.chars().take(200).collect::<String>()
        ));
    }

    let bytes = resp
        .bytes()
        .await
        .map_err(|e| format!("Failed to read response: {}", e))?;

    tokio::fs::write(dest, &bytes)
        .await
        .map_err(|e| format!("Failed to write file: {}", e))?;

    Ok(())
}

pub fn extract_mmdb_from_tar_gz(
    tar_gz_path: &Path,
    edition: &str,
    output_path: &Path,
) -> Result<(), String> {
    use flate2::read::GzDecoder;
    use std::fs::File;
    use tar::Archive;

    let file = File::open(tar_gz_path).map_err(|e| format!("Failed to open tar.gz: {}", e))?;
    let decoder = GzDecoder::new(file);
    let mut archive = Archive::new(decoder);

    let mmdb_filename = format!("{}.mmdb", edition);

    for entry in archive
        .entries()
        .map_err(|e| format!("Failed to read tar entries: {}", e))?
    {
        let mut entry = entry.map_err(|e| format!("Failed to read tar entry: {}", e))?;
        let path = entry
            .path()
            .map_err(|e| format!("Failed to get entry path: {}", e))?;

        if path
            .file_name()
            .and_then(|f| f.to_str())
            .map_or(false, |name| name == mmdb_filename)
        {
            let mut output_file = std::fs::File::create(output_path)
                .map_err(|e| format!("Failed to create output file: {}", e))?;
            std::io::copy(&mut entry, &mut output_file)
                .map_err(|e| format!("Failed to extract file: {}", e))?;
            return Ok(());
        }
    }

    Err(format!("{} not found in downloaded archive", mmdb_filename))
}

async fn download_and_extract(
    edition: &str,
    license_key: &str,
    db_dir: &Path,
    app: &AppHandle,
) -> Result<PathBuf, String> {
    let url = build_download_url(edition, license_key);
    log::info!("[geodb] Downloading {}", edition);

    tokio::fs::create_dir_all(db_dir)
        .await
        .map_err(|e| format!("Failed to create database directory: {}", e))?;

    emit_progress(
        app,
        &GeoDbDownloadProgress {
            stage: "downloading".to_string(),
            edition: edition.to_string(),
            percent: Some(0),
            error: None,
        },
    );

    let tar_gz_path = db_dir.join(format!("{}.tar.gz", edition));
    download_file(&url, &tar_gz_path).await?;

    emit_progress(
        app,
        &GeoDbDownloadProgress {
            stage: "downloading".to_string(),
            edition: edition.to_string(),
            percent: Some(100),
            error: None,
        },
    );

    emit_progress(
        app,
        &GeoDbDownloadProgress {
            stage: "extracting".to_string(),
            edition: edition.to_string(),
            percent: None,
            error: None,
        },
    );

    let mmdb_path = db_dir.join(format!("{}.mmdb", edition));
    let tar_gz_clone = tar_gz_path.clone();
    let mmdb_clone = mmdb_path.clone();
    let edition_str = edition.to_string();

    tokio::task::spawn_blocking(move || {
        extract_mmdb_from_tar_gz(&tar_gz_clone, &edition_str, &mmdb_clone)
    })
    .await
    .map_err(|e| format!("Task join error: {}", e))??;

    // Clean up tar.gz
    let _ = tokio::fs::remove_file(&tar_gz_path).await;

    emit_progress(
        app,
        &GeoDbDownloadProgress {
            stage: "complete".to_string(),
            edition: edition.to_string(),
            percent: Some(100),
            error: None,
        },
    );

    Ok(mmdb_path)
}

pub async fn download_geo_databases(
    license_key: &str,
    app_data_dir: &Path,
    app: &AppHandle,
    store: &SharedStore,
    readers: &SharedGeoReaders,
) -> Result<(String, String), String> {
    let db_dir = app_data_dir.join("databases");
    let mut results: std::collections::HashMap<String, PathBuf> = std::collections::HashMap::new();

    for edition in EDITIONS {
        match download_and_extract(edition, license_key, &db_dir, app).await {
            Ok(path) => {
                log::info!("Downloaded {} to {:?}", edition, path);
                results.insert(edition.to_string(), path);
            }
            Err(e) => {
                log::error!("Failed to download {}: {}", edition, e);
                emit_progress(
                    app,
                    &GeoDbDownloadProgress {
                        stage: "error".to_string(),
                        edition: edition.to_string(),
                        percent: None,
                        error: Some(e.clone()),
                    },
                );
                return Err(e);
            }
        }
    }

    let city_path = results
        .get("GeoLite2-City")
        .ok_or("City database not downloaded")?
        .to_string_lossy()
        .to_string();
    let asn_path = results
        .get("GeoLite2-ASN")
        .ok_or("ASN database not downloaded")?
        .to_string_lossy()
        .to_string();

    // Update settings
    {
        let mut guard = store.lock().unwrap();
        guard.geo.city_db_path = Some(city_path.clone());
        guard.geo.asn_db_path = Some(asn_path.clone());
        guard.geo.last_updated = Some(chrono::Utc::now().timestamp_millis() as f64);
    }

    reload_geo_databases(readers, store);

    Ok((city_path, asn_path))
}
